File: src/lib/stats/text.ts
const URL_RE = /https?:\/\/[^\s<>)"']+/g;
const EN_WORD_RE = /[A-Za-z][A-Za-z0-9_'-]*/g;
const CJK_RE = /[\u3400-\u4dbf\u4e00-\u9fff]+/g;
const DIGIT_RE = /\d/g;

const EN_STOPWORDS = new Set([
  "a", "an", "and", "are", "as", "at", "be", "by", "for", "from",
  "has", "have", "i", "in", "is", "it", "of", "on", "or", "that",
  "the", "this", "to", "was", "were", "with", "you", "your",
  "class", "com", "const", "data", "div", "false", "function", "html",
  "http", "https", "id", "img", "info", "let", "name", "null", "span",
  "string", "true", "type", "var", "www",
]);

const CJK_STOP_BIGRAMS = new Set([
  "这个", "那个", "什么", "一下", "一个", "可以", "就是", "如果",
  "但是", "然后", "所以", "因为", "我们", "你们", "他们", "它们",
]);

export function visibleCharCount(text: string | null | undefined): number {
  return text ? text.replace(/\s+/g, "").length : 0;
}

export function excerpt(text: string | null | undefined, limit = 240): string | null {
  if (!text) {
    return null;
  }
  const compact = text.replace(/\s+/g, " ").trim();
  if (compact.length <= limit) {
    return compact;
  }
  return `${compact.slice(0, Math.max(0, limit - 1))}...`;
}

export function urlCount(text: string | null | undefined): number {
  return text ? [...text.matchAll(URL_RE)].length : 0;
}

export function codeBlockCount(text: string | null | undefined): number {
  if (!text) {
    return 0;
  }
  const fenced = Math.floor((text.match(/```/g) || []).length / 2);
  const inlineLike = (text.match(/`[^`\n]{3,}`/g) || []).length;
  return fenced + inlineLike;
}

export function punctuationCounts(text: string | null | undefined): Record<string, number> {
  if (!text) {
    return { question_marks: 0, exclamation_marks: 0 };
  }
  return {
    question_marks: countChars(text, "?") + countChars(text, "？"),
    exclamation_marks: countChars(text, "!") + countChars(text, "！"),
  };
}

export function languageCounts(text: string | null | undefined): Record<string, number> {
  if (!text) {
    return { chinese_characters: 0, english_words: 0, digit_characters: 0 };
  }
  return {
    chinese_characters: [...text.matchAll(CJK_RE)].reduce((total, match) => total + match[0].length, 0),
    english_words: [...text.matchAll(EN_WORD_RE)].length,
    digit_characters: [...text.matchAll(DIGIT_RE)].length,
  };
}

export function languageBucket(text: string | null | undefined): string {
  const counts = languageCounts(text);
  const hasCjk = counts.chinese_characters > 0;
  const hasEnglish = counts.english_words > 0;
  if (hasCjk && hasEnglish) return "mixed";
  if (hasCjk) return "chinese";
  if (hasEnglish) return "english";
  return "other";
}

export function frequentTerms(texts: Iterable<string | null | undefined>, limit = 50): Array<{ term: string; count: number }> {
  const counter = new Map<string, number>();
  for (const text of texts) {
    if (!text) {
      continue;
    }
    englishTerms(text).forEach((term) => increment(counter, term));
    cjkBigrams(text).forEach((term) => increment(counter, term));
  }
  return [...counter.entries()]
    .sort((left, right) => right[1] - left[1] || left[0].localeCompare(right[0]))
    .slice(0, limit)
    .map(([term, count]) => ({ term, count }));
}

function englishTerms(text: string): string[] {
  const terms: string[] = [];
  for (const match of text.matchAll(EN_WORD_RE)) {
    const term = match[0].toLowerCase().replace(/^['_-]+|['_-]+$/g, "");
    if (term.length >= 2 && !EN_STOPWORDS.has(term)) {
      terms.push(term);
    }
  }
  return terms;
}

function cjkBigrams(text: string): string[] {
  const terms: string[] = [];
  for (const match of text.matchAll(CJK_RE)) {
    const seq = match[0];
    if (seq.length === 2 && !CJK_STOP_BIGRAMS.has(seq)) {
      terms.push(seq);
    } else if (seq.length > 2) {
      for (let index = 0; index < seq.length - 1; index += 1) {
        const term = seq.slice(index, index + 2);
        if (!CJK_STOP_BIGRAMS.has(term)) {
          terms.push(term);
        }
      }
    }
  }
  return terms;
}

function countChars(text: string, char: string): number {
  return text.split(char).length - 1;
}

function increment(counter: Map<string, number>, key: string, amount = 1): void {
  counter.set(key, (counter.get(key) || 0) + amount);
}

File: chatgpt_wrapped/text.py
from __future__ import annotations

import re
from collections import Counter
from typing import Iterable


URL_RE = re.compile(r"https?://[^\s<>)\"']+")
EN_WORD_RE = re.compile(r"[A-Za-z][A-Za-z0-9_'-]*")
CJK_RE = re.compile(r"[\u3400-\u4dbf\u4e00-\u9fff]+")
DIGIT_RE = re.compile(r"\d")

EN_STOPWORDS = {
    "a", "an", "and", "are", "as", "at", "be", "by", "for", "from",
    "has", "have", "i", "in", "is", "it", "of", "on", "or", "that",
    "the", "this", "to", "was", "were", "with", "you", "your",
    "class", "com", "const", "data", "div", "false", "function", "html",
    "http", "https", "id", "img", "info", "let", "name", "null", "span",
    "string", "true", "type", "var", "www",
}

CJK_STOP_BIGRAMS = {
    "这个", "那个", "什么", "一下", "一个", "可以", "就是", "如果",
    "但是", "然后", "所以", "因为", "我们", "你们", "他们", "它们",
}


def visible_char_count(text: str | None) -> int:
    if not text:
        return 0
    return len(re.sub(r"\s+", "", text))


def excerpt(text: str | None, limit: int = 240) -> str | None:
    if not text:
        return None
    compact = re.sub(r"\s+", " ", text).strip()
    if len(compact) <= limit:
        return compact
    return compact[: max(0, limit - 1)] + "…"


def url_count(text: str | None) -> int:
    if not text:
        return 0
    return len(URL_RE.findall(text))


def code_block_count(text: str | None) -> int:
    if not text:
        return 0
    fenced = text.count("```") // 2
    inline_like = len(re.findall(r"`[^`\n]{3,}`", text))
    return fenced + inline_like


def punctuation_counts(text: str | None) -> dict[str, int]:
    if not text:
        return {"question_marks": 0, "exclamation_marks": 0}
    return {
        "question_marks": text.count("?") + text.count("？"),
        "exclamation_marks": text.count("!") + text.count("！"),
    }


def language_counts(text: str | None) -> dict[str, int]:
    if not text:
        return {"chinese_characters": 0, "english_words": 0, "digit_characters": 0}
    return {
        "chinese_characters": sum(len(match.group(0)) for match in CJK_RE.finditer(text)),
        "english_words": len(EN_WORD_RE.findall(text)),
        "digit_characters": len(DIGIT_RE.findall(text)),
    }


def language_bucket(text: str | None) -> str:
    counts = language_counts(text)
    has_cjk = counts["chinese_characters"] > 0
    has_en = counts["english_words"] > 0
    if has_cjk and has_en:
        return "mixed"
    if has_cjk:
        return "chinese"
    if has_en:
        return "english"
    return "other"


def frequent_terms(texts: Iterable[str | None], *, limit: int = 50) -> list[dict[str, int | str]]:
    counter: Counter[str] = Counter()
    for text in texts:
        if not text:
            continue
        counter.update(_english_terms(text))
        counter.update(_cjk_bigrams(text))
    return [
        {"term": term, "count": count}
        for term, count in counter.most_common(limit)
    ]


def _english_terms(text: str) -> list[str]:
    terms = []
    for match in EN_WORD_RE.finditer(text):
        term = match.group(0).lower().strip("'_-")
        if len(term) >= 2 and term not in EN_STOPWORDS:
            terms.append(term)
    return terms


def _cjk_bigrams(text: str) -> list[str]:
    terms: list[str] = []
    for match in CJK_RE.finditer(text):
        seq = match.group(0)
        if len(seq) == 2 and seq not in CJK_STOP_BIGRAMS:
            terms.append(seq)
        elif len(seq) > 2:
            for i in range(len(seq) - 1):
                term = seq[i : i + 2]
                if term not in CJK_STOP_BIGRAMS:
                    terms.append(term)
    return terms
